docs/developers/Custom_Rules.mdx

```diff
@@ -272,6 +272,12 @@ Rules can retrieve their full backing TypeScript type checker with `services.pro
 This can be necessary for TypeScript APIs not wrapped by the parser services.
 :::
 
+:::caution
+We recommend against changing rule logic on whether `services.program` exists.
+In our experience, users are generally surprised when rules behave differently with or without type information.
+Additionally, if they misconfigure their ESLint config, they may not realize why the rule started behaving differently.
```

Comment:
This sounds vague. Maybe a specific case?

Reply:
Hmm yeah good point. Agreed.

Reply:
as possible solution we could recommend to create 2 separate rules one with type-info and one without

Reply:
I like it! Added :)

Reply:
Also gating it behind options is *okay*, I think.
That way you still have that hard point of a user explicitly opting in to the type-awareness and thus the rule can explicitly crash if it's not provided.

`naming-convention` does this (the only rule we have that does) for the type-based enforcement of names.

Reply:
This actually is very relevant to https://github.com/veritem/eslint-plugin-vitest/issues/251. Nice.